### ISSUE TITLE
Nodogsplash not working without internet

### ISSUE THREAD

**Issue by ajaymandera** [nodogsplash/nodogsplash#154]
I have successfully installed NoDogsplash captive portal on ubuntu 16.04 and it's working fine when the device is connected to LAN. Now the issue is when the device is not connected to LAN, NoDogSplash is not working and devices are not able to connect to AP or if they made a successful connection then redirection doesn't work. 
Also noticed that while not connected to LAN, NoDogsplash is eating 100% CPU.

**Comment by mwarning** [comment]
What ndg version do you use?

**Comment by bluewavenet** [comment]
@ajaymandera , can we assume you mean WAN rather than LAN?
NDS requires an Internet connection to work in normal circumstances as clients attaching to an AP will run their inbuilt CPD (Captive Portal Detection) and attempt to contact a set http port 80 address (determined by the operating system of the client device). The first thing the client's CPD will do after the client has received its DHCP IP address is to do a DNS query. If the Internet connection(WAN) is not available, then this will fail and the CPD will not issue its port 80 request and in turn NDS will not receive that request or serve its splash page.

In summary, what you are seeing is expected.

As for the 100% CPU, I have not seen this so cannot comment.

**Comment by ajaymandera** [comment]
@mwarning  ndg version is 1.0.0

@bluewavenet no, I am talking about LAN(the incoming internet connection to my device, here PI3).
When LAN is up also WAN is up and Ndg functioning smoothly. But the problem arises when LAN is down.
Can anyone suggest me what to do in this case? I want to setup small internal network through PI with the help of ndg captive portal. 
Internet availability is not stable on PI so I want ndg to work without internet as well, is it possible? anyway I can try?

**Comment by bluewavenet** [comment]
Hi @ajaymandera 
The normal terminology is the Internet feed to NDS will be its WAN and the connection to it's clients will be LAN.
It is easy to make an offline captive web site, for use in a museum for example. All you have to do is add the line:
address=/#/[ip address of your museum tour webserver]
to the config of dnsmasq.
This however will stop any access to the Internet. It should be possible to use this along with a script to detect if the Internet feed is up and modify dnsmasq config and restart it automatically.
I hope this helps ;)

**Comment by bluewavenet** [comment]
@lynxis 
This will apply to version 2 as well, I think. To automate, It is more of an enhancement to Dnsmasq than NDS but would be nice to have at some stage - pretty low priority though ;)

**Comment by zaolin** [comment]
@bluewavenet I will push a pull request soon which solves this issue

**Comment by nicopace** [comment]
@ajaymandera I managed to use nodogsplash without internet by capturing all dns requests and redirecting them to a spoofed dns server that always replies with the gateway ip.
This only applies to dns requests from unauthorized clients.

You can check the hack here:
https://chef.altermundi.net/diff/laotrared-node/libre-mesh-1509/

Regards,

**Comment by kerematam** [comment]
This feature is needed as default configuration. Because on my tests so far; in case of slow WAN or WLAN connection Nodogsplah needs to wait for DNS requests to be returned... After that, captive portal intercepts. So Nodogsplash unnecessarily waits for DNS messaging to forward clients. And as i noticed some of WiFi driver of client devices fall into timeout and does not redirect.

@nicopace good approach. but i couldnt understand your patch. it seem like you only removed some of init files and add nothing?

**Comment by nicopace** [comment]
I added this as an init script

```
/etc/init.d/nodogsplash stop

iptables -t nat -N dnspreauth

iptables -t nat -A PREROUTING -p udp --dport 53 -j dnspreauth
iptables -t nat -A PREROUTING -p tcp --dport 53 -j dnspreauth

iptables -t nat -A dnspreauth -m mark --mark 0x100/0x700 -j RETURN
iptables -t nat -A dnspreauth -p udp --dport 53 -j DNAT  --to $(uci get network.lan.ipaddr):553
iptables -t nat -A dnspreauth -p tcp --dport 53 -j DNAT  --to $(uci get network.lan.ipaddr):553
iptables -t nat -A dnspreauth -j RETURN

/etc/init.d/nodogsplash start
```

And also started a second dnsmasq with this config
```
-port=553
```

You need to add something else in the dnsmasq config, but don't remember how.

**Comment by gregorysemah** [comment]
Hello guys and thanks for your work.
I have installed nodogsplash on routeur TL-WR841N with OpenWrt firmware
I'am trying to achieve a captive portal with no internet connection (like for the museum example above).
@bluewavenet by ssh login to the router, adding this famous line which could resolve everything for my case:
address=/#/192.168.1.242 (ip of my webserver)
It is not working, can you make little more explanation of what i have to do to make it work to resolve this issue?

**Comment by tim-moody** [comment]
I am also very interested in this functionality, but I have not managed to get any of the suggestions above to work. See also my comments to #177 . I work on a project called Internet in a Box, which provides internet content, such as wikis, in an offline environment for schools and medical facilities in developing countries (https://github.com/iiab/iiab) and we really need captive portal to provide 'museum' mode access to the home page.

**Comment by bluewavenet** [comment]
@tim-moody 
I have had this working with the Dnsmasq config change quite some time ago and with a very old version of NDS. I will give it a quick try with a current version.
One of the big problems with this is the captive portal detection (CPD) now present on most devices in one form or another. The latest incarnations of CPD do the following:
1. On connection to a wifi, the Network Manager invokes CPD
2. CPD tries to connect to a pre determined port 80 web page, with the knowledge of the specific response that page will return. It continues to poll this page at a set interval until a response is received or the WiFi is disconnected. This interval is often 300 seconds but can vary from vendor to vendor.
3. CPD checks the returned response. If it is the correct response, then we have an Internet connection and CPD terminates.
4. If we have an incorrect response, this will be a Captive Portal (CP) responding with its splash page so the CPD opens a "browser" to display that page and pipes the page data to it. CPD signals the Network Manager that a CP has been detected so the wifi/? symbol or message to user etc can be displayed.
5. The client interacts with the splash page to log in and get Internet access one way or another (eg create a free account, pay for use or just accept terms of service).
6. CPD detects external network traffic so retries its pre determined port 80 page and if it receives the expected response, it informs Network Manager (so the WiFi "?" can be removed etc), terminates the browser it opened and exits.

Not all implementations of CPD do this in this exact way, but the principle is the same.

Your problem will be the "browser" the CPD pops up. This could be a very stripped down built in, or a restricted invocation of a browser app, depending on vendor implementation and/or what the client device has installed, all things we have no control over.
Most CPD browser invocations will prohibit file downloads and many have javascript disabled, both for security reasons.

I hope this clarifies things a bit.
I will do that quick test with a current version of NDS.

**Comment by bluewavenet** [comment]
@tim-moody 
I can confirm the dnsmasq config change works on the test system I have here.
It just happens to be NDS v1.0.1 running on LEDE/OpenWrt 17.01.4
Should work with NDS 2.1.1 with no problems.

**Comment by tim-moody** [comment]
To be clear. I need this to work with no internet connection. I am not trying to set up an internet cafe or authenticate users before giving them access to the internet. I have a completely offline environment in which I simply want to automate access to the server home page (address=/#/<my ip> is not sufficient). I am also on a Raspberry Pi, not a router. In short item 6 above is not expected.

My test is dead simple. Power on the rpi with 2.x nds installed and running with a wired internet connection and hostapd providing the hotspot. Connect with a win10 laptop and a recent android phone. Both bring up the splash page, which can be clicked to reach the server home page, so the problem is not likely the browser.

If I power off the rpi, remove the ethernet connection to the internet, and power on, both devices can connect to the hotspot, but the splash page never appears. If I apply @nicopace 's iptables rules and set up a separate dnsmasq on 553 the situation doesn't change.

Perhaps @bluewavenet or someone could give me a dnsmasq.conf file that works. I took mine from @zaolin fixing the syntax problems, but it did not work. (Yes I know he used 5353 instead of 553.)

Any help appreciated.

**Comment by bluewavenet** [comment]
@tim-moody 
I think perhaps you are over thinking this.
Assuming you do not normally (or ever) need Internet access, then to make this work you do not need any of @zaolin 's patches. Those patches were to automate a changeover in the case of the Internet feed being down in a normal CP situation.

Also NDS must run on a device _**configured**_ to be a router that must provide the default gateway and dhcp for its lan.

The _**single**_ line required in the dnsmasq configuration is as follows:
 address=/#/[lan_ip_address_of_NDS]
eg in my test system working here now:
 address=/#/10.168.1.1

The CPD functionality I described previously is what _**WILL**_ almost always happen as all mobile devices and most other devices and browsers now include a CPD function.

If you design your splash page such that it does not authenticate to NDS, then the sequence will not get to item 6.

With no internet connection and the added line in the dnsmasq config, the CPD on the device will still pop up the CPD browser. I have just tested again with an android phone, a linux laptop and a windows 10 laptop and it works with all of them without an Internet connection.

The only issue is the CPD browser may or **_may not_** allow downloads, allow links, allow javascript (depending on the operating system, vendor, apps installed etc etc).

**Comment by bluewavenet** [comment]
@tim-moody 
You might find #249 relevant.

**Comment by bluewavenet** [comment]
For offline systems the dnsmasq config works very well ie:

> The single line required in the dnsmasq configuration is as follows:
address=/#/[lan_ip_address_of_NDS]
eg in my test system working here now:
address=/#/10.168.1.1

If there is a demand for an automated process to switch between online and offline modes then this can be re-opened.

**Comment by charlyjpc** [comment]
it's a samsung problem
https://community.arubanetworks.com/t5/Wireless-Access/Samsung-Captive-Portal-Detection/td-p/404950

**Comment by bluewavenet** [comment]
@charlyjpc 
Sorry but no, it is not a Samsung problem.
It is the client device Captive Portal Detection working as designed. This is a built in function of all major operating systems and browsers and is a defacto industry standard.

**Comment by bomboem** [comment]
J

**Comment by thormj** [comment]
So... what's the right way to request to reopen this?
I think it would be most excellent to work when offline and transition to online because that would handle intermittent connections by default.  Eg,

If the NoDogSplash redirect could supply indicators to the "splash" page (eg, "DNS not working", "No Traffic Working", "Interface Down"), it would be awesome (yay!  debugging!).

My particular application is like the museum bit (on a RasPi0 -- no storage!), but if the internet is working, I'd like to display things like "the making of" or "the history behind."  -- When you connect, the CPD pops up a basic "about this exhibit" and then the "login/authenticate" is a button "Would You Like To Learn More..." that redirects you to the "full page" (over on Rackspace).

The CPD is great because --as soon as you get within range (turned down rf power on the pi), the CPD portal pops up and says "you're looking at this artifact" (some phones even let you do audio!) ... and they all have unique "world" URLS (no "just" having everything be /# and hoping the browser figures it out). -- You have a QR code for the hotspot name and "automatically connect."

**Comment by bluewavenet** [comment]
@thormj 
Be aware that NodogSplash has split into two projects:
 1. NoDogSplash - stripped down and FAS removed to optimise for legacy devices with very restricted resources.
 2. openNDS - containing the full FAS API for developing custom dynamic splash pages.

As "no code" is the "smallest code", the simple Dnsmasq hack is the best way forward for NoDogSplash.

I suggest you open a feature request (as a new issue) at https://github.com/openNDS/openNDS

**Comment by alpha77788998778999877855** [comment]
Hi @bluewavenet , 
`As "no code" is the "smallest code", the simple Dnsmasq hack is the best way forward for NoDogSplash.`
I have a raspberry which has a QR code on its screen, when I take a picture of the QR code, nothing happens, no opening of the web browser. I have several localhost sites with a back and a front and mysql and a wifi hotspot for participants to participate on various web applications with their phone or their tablet or their computer. I installed nodogsplash to create a quick connection without having to enter the site url and go directly to an application portal. I have a static ip address in the dnsmasq.conf. I tried a lot of things ...

**Comment by bluewavenet** [comment]
@alpha77788998778999877855 
If you are trying to have a "Museum" type system with a pi screen with QR code at each exhibit, you really need openNDS/FAS running in offline mode. If this is the case, please open an issue there.
https://github.com/openNDS/openNDS
